[HZ-569] use ConfigValidator consistently

alex-dukhno opened the pull request:
Fix for #16009 

Regarding the [comment](https://github.com/hazelcast/hazelcast/pull/15939#discussion_r345770523)
> The method ConfigValidator#checkMapEvictionConfig(com.hazelcast.config.EvictionConfig) is used from the ConfigValidator#` but also from the MemberDomConfigProcessor#doEvictionConfigChecks. The first method is invoked when a distributed map is created but the second one is invoked during parse time. Can the config parser use the checkMapConfig too so the checkMapEvictionConfig becomes private?
1) I make `checkMapEvictionConfig` package protected because it is used in `ConfigValidatorEvictionConfigTest#checkEvictionConfig_forMap` test.
2) I feel like that changes aren't worth it ... trying to replace `checkMapEvictionConfig` call with `checkMapConfig` has a ripple effect. While you want to make the method `private` you have to make others `public` and call from other classes

Anyway, @mmedenjak, @ahmetmircik could you guys take a look, maybe you will have a better idea what can be done here?

Thank you

devOpsHazelcast commented:
Can one of the admins verify this patch?

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/internal/config/ConfigValidator.java:
```diff
@@ -128,16 +124,16 @@ public static void checkMapConfig(MapConfig mapConfig,
                                       SplitBrainMergePolicyProvider mergePolicyProvider,
                                       HazelcastProperties properties) {
 
-        checkNotNativeWhenOpenSource(mapConfig.getInMemoryFormat());
```
How come we don't check this anymore when the config is parsed through `MemberDomConfigProcessor`?

alex-dukhno replied:
There was a test failure, but I could not remember. I tried run `ConfigValidatorTest`, the only test class that is affected and it is green.
I'll revert the change

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/internal/config/MemberDomConfigProcessor.java:
```diff
@@ -1788,7 +1787,13 @@ void handleMapNode(Node parentNode, final MapConfig mapConfig) throws Exception
                 mapConfig.setAsyncBackupCount(getIntegerValue("async-backup-count", value));
             } else if ("eviction".equals(nodeName)) {
-                mapConfig.setEvictionConfig(getEvictionConfig(node, false, true));
+                EvictionConfig evictionConfig = getEvictionConfig(node, false, true);
```
I was actually thinking if it would not make sense that we validate the whole map config, just like we do when creating a new distributed object, like this:
```
    void handleMapNode(Node parentNode, final MapConfig mapConfig) throws Exception {
        for (Node node : childElements(parentNode)) {
            ...
        }
        try {
            checkMapConfig(mapConfig, config.getNativeMemoryConfig(), null, new HazelcastProperties(config));
        } catch (IllegalArgumentException e) {
            throw new InvalidConfigurationException(e.getMessage());
        }
        config.addMapConfig(mapConfig);
    }
```
Not sure if there's anything preventing us from doing so cc @ahmetmircik .

alex-dukhno replied:
@mmedenjak do you mean put down `try` `catch` block at the end of the `handleMapNode` method?

mmedenjak replied:
Yes, I was thinking if we can (and if we can't, why) have just one method exposed on `ConfigValidator` that validates the entire map config and use it whenever the whole map config is either parsed or when we go to create a new distributed object.

alex-dukhno replied:
I've updated the code according to the discussion

mmedenjak commented:
run-lab-run

mmedenjak commented:
FYI, these were the failed tests, you can try them out locally:
```
com.hazelcast.config.ConfigXmlGeneratorTest.testAttributesConfigWithStoreFactory
com.hazelcast.config.ConfigXmlGeneratorTest.testAttributesConfigWithStoreClass
com.hazelcast.config.XMLConfigBuilderTest.testMapNativeMaxSizePolicy
com.hazelcast.config.YamlConfigBuilderTest.testMapNativeMaxSizePolicy
```

alex-dukhno commented:
@mmedenjak I've updated tests. That happened because of `BINARY` in memory formats and `checkNotNativeWhenOpenSource` was failing.

Thank you

mmedenjak commented:
run-lab-run

mmedenjak commented:
Hi, now it's these two, they look simple to fix:
```
com.hazelcast.config.XMLConfigBuilderTest.testMapNativeMaxSizePolicy
com.hazelcast.config.YamlConfigBuilderTest.testMapNativeMaxSizePolicy
```

alex-dukhno commented:
Hi @mmedenjak, I've fixed the tests, but the problem is that tests say `testMapNativeMaxSizePolicy` and I change `NATIVE` format to `OBJECT`.
Thus we either should rename tests or move them to non open source solution because of `checkNotNativeWhenOpenSource` ... or probably both?

WDYT?

mmedenjak commented:
@alex-dukhno hm, yes, maybe both. You can rename the test and we'll add the enterprise version in our private repo once this is merged.

alex-dukhno commented:
@mmedenjak done

mmedenjak commented:
run-lab-run

alex-dukhno commented on hazelcast/src/main/java/com/hazelcast/internal/config/MemberDomConfigProcessor.java:
```diff
@@ -1837,6 +1836,11 @@ void handleMapNode(Node parentNode, final MapConfig mapConfig) throws Exception
                 mapQueryCacheHandler(node, mapConfig);
             }
         }
+        try {
```
I've never worked with Spirng `BeanDefinition` programmatically. Can any one help me out how can we validate `MapConfig` bean definition?

mmedenjak replied:
Hm, seems I was wrong, there is no validation there, probably because of the `BeanDefinition` usage. Ignore this then.

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/internal/config/MemberDomConfigProcessor.java:
```diff
@@ -2048,32 +2052,9 @@ private EvictionConfig getEvictionConfig(Node node, boolean isNearCache, boolean
             evictionConfig.setComparatorClassName(getTextContent(comparatorClassName));
         }
-
-        try {
```
@alex-dukhno have you been able to take a look at this?

ahmetmircik replied:
@mmedenjak we need to validate query-cache eviction config as others. 
Seems ok to do it like this:
```
 for (QueryCacheConfig queryCacheConfig : mapConfig.getQueryCacheConfigs()) {
            checkCacheEvictionConfig(queryCacheConfig.getEvictionConfig());
        }
```
With the goal of decreasing complexity, other things you mentioned makes sense to me.

alex-dukhno replied:
@mmedenjak I have read it, but decided to take a look after #16157

alex-dukhno commented:
@mmedenjak I've updated code as per your https://github.com/hazelcast/hazelcast/pull/16095#discussion_r355929405
Things to notice:
1) I renamed the tests [here](https://github.com/hazelcast/hazelcast/pull/16095/files#diff-d86a9636d620f9a4e8499848b7158c5fR185). The issue is the same as with in the https://github.com/hazelcast/hazelcast/pull/16095#issuecomment-562608052
2) when I added `initDefaultMaxSizeForOnHeapMaps(mapConfig.getNearCacheConfig());` line [here](https://github.com/hazelcast/hazelcast/pull/16095/files#diff-98ddc144064f4ff98815e8fcd855effdR1780) the `XmlYamlConfigBuilderEqualsTest#testFullConfigNormalNetwork` and `XmlYamlConfigBuilderEqualsTest#testFullConfigAdvancedNetwork` on different eviction sizes for default map's near-cache. For xml it was `10_000` for yaml it was `Integer#MAX_VALUE`. I couldn't find code that would be different for parsing xml and yaml that would have such effect.

Thank you

ahmetmircik commented:
run-lab-run

ahmetmircik commented:
2 Checkstyle issues:
```
13:29:15 [INFO] Starting audit...
13:29:15 [ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-
builder_2/hazelcast/src/main/java/com/hazelcast/internal/config/MemberDomConfigProcessor.java:154:15: Unused import - 
com.hazelcast.config.NearCacheConfigAccessor.initDefaultMaxSizeForOnHeapMaps. 
[UnusedImports]
13:29:15 [ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/config/MemberDomConfigPro
cessor.java:201: Using the '.*' form of import should be avoided - 
com.hazelcast.internal.config.ConfigValidator.*.
[AvoidStarImport]
13:29:15 Audit done.
```

alex-dukhno commented:
I've pushed check style fix

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

ahmetmircik commented:
run-lab-run

ahmetmircik commented:
5 failed tests:
```
Test Result (5 failures /+5)
com.hazelcast.config.ConfigXmlGeneratorTest.testMapNearCacheConfig
com.hazelcast.map.impl.mapstore.EntryLoaderDistributedTest.testEntrySet_withPredic
ate[inMemoryFormat: BINARY]
com.hazelcast.map.impl.mapstore.EntryLoaderDistributedTest.testEntrySet[inMemoryFo
rmat: BINARY]
com.hazelcast.map.impl.mapstore.writebehind.EntryStoreWriteBehindDistributedTest.t
estSetAsync_withMaxIdle[inMemoryFormat: OBJECT]
com.hazelcast.map.impl.proxy.MapProxyImplTest.whenNearCachedMapProxyIsCreat
ed_mapContainerIsNotCreated
```

alex-dukhno commented:
Hi @ahmetmircik, I could only reproduce failure of the `com.hazelcast.config.ConfigXmlGeneratorTest.testMapNearCacheConfig` test locally. 

Can you, please, post error messages from other tests that I could fix them?

Thank you

ahmetmircik commented:
run-lab-run

ahmetmircik commented:
run-lab-run

devOpsHazelcast commented:
[![CLA assistant check](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/pull/badge/not_signed)](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast?pullRequest=16095) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast?pullRequest=16095) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/check/hazelcast/hazelcast?pullRequest=16095) it.</sub>

pveentjer commented:
Closing PR due to lack of progress.
